libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc

```diff
@@ -0,0 +1,528 @@
+template<typename FieldT, size_t generator>
+flystel_Q_delta_binary_field_gadget<FieldT, generator>::
+    flystel_Q_delta_binary_field_gadget(
+        protoboard<FieldT> &pb,
+        const linear_combination<FieldT> &input,
+        const pb_variable<FieldT> &output,
+        const std::string &annotation_prefix)
+    : gadget<FieldT>(pb, annotation_prefix)
+    , internal(pb_variable_allocate<FieldT>(
+          pb, FMT(this->annotation_prefix, " internal")))
+#ifdef FLYSTEL_DEBUG
+    , beta(DEBUG_FLYSTEL_BETA)
+    , delta(DEBUG_FLYSTEL_DELTA)
+#elif
+    , beta(FieldT(generator))
+    , delta(FieldT(0))
+#endif // #ifdef FLYSTEL_DEBUG
+    , input(input)
+    , output(output)
+{
+}
+
+template<typename FieldT, size_t generator>
+void flystel_Q_delta_binary_field_gadget<FieldT, generator>::
+    generate_r1cs_constraints()
+{
+    // (beta * input) * input = internal
+    this->pb.add_r1cs_constraint(
+        r1cs_constraint<FieldT>(beta * input, input, internal),
+        FMT(this->annotation_prefix, " beta * x * x = x_square"));
+    // internal * input = output - delta
+    this->pb.add_r1cs_constraint(
+        r1cs_constraint<FieldT>(internal, input, output - delta),
+        FMT(this->annotation_prefix, " x_square * x = y - delta"));
+}
+
+template<typename FieldT, size_t generator>
+void flystel_Q_delta_binary_field_gadget<FieldT, generator>::
+    generate_r1cs_witness()
+{
+    const FieldT input_value =
+        input.evaluate(this->pb.full_variable_assignment());
+    // x_internal = beta x * x
+    this->pb.val(internal) = (this->beta * input_value) * input_value;
+    // y = beta x^3 + delta = x_internal * x + delta
+    this->pb.val(output) = this->pb.val(internal) * input_value + this->delta;
+
+    printf("[%s:%d] %s() input ", __FILE__, __LINE__, __FUNCTION__);
+    input_value.print();
+    printf("[%s:%d] %s() output ", __FILE__, __LINE__, __FUNCTION__);
+    this->pb.val(output).print();
+}
+
+// R1CS constraints for the operation y = x^5 with x=input,
+// y=output. This operation is represented with three multiplications
+// as y = (temp * temp * x), temp = x * x. Equivalently:
+//
+// x  * x  = a0
+// a0 * a0 = a1
+// x  * a1 =  y
+//
+// for the variables vector X = (x0=1, x1=x, x2=a0, x3=a1, x4=y). The
+// above system is represented with 3 R1CS constraints resp.:
+//
+// < A0 , X > * < B0 , X > = < C0 , X > ,
+// < A1 , X > * < B1 , X > = < C1 , X > ,
+// < A2 , X > * < B2 , X > = < C2 , X >
+//
+// where A0=(01000), B0=(01000), C0=(00100); A1=(00100), B0=(00100),
+// C0=(00010) and A2=(01000), B2=(00010), C2=(00001)
+template<typename FieldT>
+flystel_E_power_five_gadget<FieldT>::flystel_E_power_five_gadget(
+    protoboard<FieldT> &pb,
+    const linear_combination<FieldT> &input,
+    const pb_variable<FieldT> &output,
+    const std::string &annotation_prefix)
+    : gadget<FieldT>(pb, annotation_prefix), input(input), output(output)
+{
+    a0.allocate(this->pb, " internal value a0");
```

Comment:
Use `pb_variable_allocate` so these are in the initializer ists and can be made const?

Reply:
Done.